swarm/src/handler/from_fn.rs

```diff
@@ -0,0 +1,349 @@
+use std::collections::VecDeque;
+use std::error::Error;
+use std::fmt;
+use std::future::Future;
+use std::task::{Context, Poll, Waker};
+use void::Void;
+
+/// A low-level building block for protocols that can be expressed as async functions.
+///
+/// An async function in Rust is executed within an executor and thus only has access to its local state
+/// and by extension, the state that was available when the [`Future`] was constructed.
+///
+/// This [`ConnectionHandler`] aims to reduce the boilerplate for protocols which can be expressed as
+/// a static sequence of reads and writes to a socket where the response can be generated with limited
+/// "local" knowledge or in other words, the local state within the [`Future`].
+///
+/// For outbound substreams, arbitrary data can be supplied via [`InEvent::NewOutbound`] which will be
+/// made available to the callback once the stream is fully negotiated.
+///
+/// Inbound substreams may be opened at any time by the remote. To facilitate this one and more usecases,
+/// the supplied callbacks for inbound and outbound substream are given access to the handler's `state`
+/// field. This `state` field can contain arbitrary data and can be updated by the [`NetworkBehaviour`]
+/// via [`InEvent::UpdateState`].
+///
+/// The design of this [`ConnectionHandler`] trades boilerplate (you don't have to write your own handler)
+/// and simplicity (small API surface) for eventual consistency, depending on your protocol design:
+///
+/// Most likely, the [`NetworkBehaviour`] is the authoritive source of `TState` but updates to it have
+/// to be manually performed via [`InEvent::UpdateState`]. Thus, the state given to newly created
+/// substreams, may be outdated and only eventually-consistent.
+///
+/// [`NetworkBehaviour`]: crate::NetworkBehaviour
+pub fn from_fn<TInbound, TOutbound, TOutboundOpenInfo, TState, TInboundFuture, TOutboundFuture>(
+    protocol: &'static str,
+    inbound_streams_limit: usize,
+    pending_dial_limit: usize,
+    on_new_inbound: impl Fn(NegotiatedSubstream, &mut TState) -> TInboundFuture + Send + 'static,
+    on_new_outbound: impl Fn(NegotiatedSubstream, &mut TState, TOutboundOpenInfo) -> TOutboundFuture
+        + Send
+        + 'static,
+) -> FromFn<TInbound, TOutbound, TOutboundOpenInfo, TState>
+where
+    TInboundFuture: Future<Output = TInbound> + Send + 'static,
+    TOutboundFuture: Future<Output = TOutbound> + Send + 'static,
+    TState: Default,
+{
+    FromFn {
+        protocol,
+        inbound_streams_limit,
+        pending_dial_limit,
+        inbound_streams: FuturesUnordered::default(),
+        outbound_streams: FuturesUnordered::default(),
+        on_new_inbound: Box::new(move |stream, state| on_new_inbound(stream, state).boxed()),
+        on_new_outbound: Box::new(move |stream, state, info| {
+            on_new_outbound(stream, state, info).boxed()
+        }),
+        idle_waker: None,
+        pending_dials: VecDeque::default(),
```

Comment:
```suggestion
        pending_outbound: VecDeque::default(),
```

It took me a bit to understand that this is an outbound stream and not connection. I associate _dial_ with a connection.

Reply:
Makes sense, I'll change it!